app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkController.kt

```diff
@@ -0,0 +1,128 @@
+    fun handleSelectionModeSwitch()
+    fun handleBookmarkEdit(node: BookmarkNode)
+    fun handleBookmarkSelected(node: BookmarkNode)
+    fun handleCopyUrl(item: BookmarkNode)
+    fun handleBookmarkSharing(item: BookmarkNode)
+    fun handleOpeningBookmark(item: BookmarkNode, mode: BrowsingModeManager.Mode)
+    fun handleBookmarkDeletion(nodes: Set<BookmarkNode>, eventType: Event)
+    fun handleBackPressed()
+    fun handleSigningIn()
+}
+
+@SuppressWarnings("TooManyFunctions")
+class DefaultBookmarkController(
+    private val context: Context,
+    private val navController: NavController,
+    private val snackbarPresenter: FenixSnackbarPresenter,
+    private val deleteBookmarkNodes: (Set<BookmarkNode>, Event) -> Unit
+) : BookmarkController {
+
+    private val activity: HomeActivity = context as HomeActivity
+    private val resources: Resources = context.resources
+    private val services: Services = activity.components.services
+
+    override fun handleBookmarkTapped(item: BookmarkNode) {
+        openInNewTab(
+            item.url!!,
+            true,
+            BrowserDirection.FromBookmarks,
+            BrowsingModeManager.Mode.Normal
+        )
+    }
+
+    override fun handleBookmarkExpand(folder: BookmarkNode) {
+        navController.nav(
+            R.id.bookmarkFragment,
+            BookmarkFragmentDirections
+                .actionBookmarkFragmentSelf(folder.guid)
+        )
+    }
+
+    override fun handleSelectionModeSwitch() {
+        activity.invalidateOptionsMenu()
+    }
+
+    override fun handleBookmarkEdit(node: BookmarkNode) {
+        navController.nav(
+            R.id.bookmarkFragment,
+            BookmarkFragmentDirections
+                .actionBookmarkFragmentToBookmarkEditFragment(node.guid)
+        )
+    }
+
+    override fun handleBookmarkSelected(node: BookmarkNode) {
+        snackbarPresenter.present(resources.getString(R.string.bookmark_cannot_edit_root))
+    }
+
+    override fun handleCopyUrl(item: BookmarkNode) {
+        item.copyUrl(context)
+        snackbarPresenter.present(resources.getString(R.string.url_copied))
+    }
+
+    override fun handleBookmarkSharing(item: BookmarkNode) {
+        navController.nav(
+            R.id.bookmarkFragment,
+            BookmarkFragmentDirections.actionBookmarkFragmentToShareFragment(
+                url = item.url!!,
+                title = item.title
+            )
+        )
+    }
+
+    override fun handleOpeningBookmark(item: BookmarkNode, mode: BrowsingModeManager.Mode) {
+        openInNewTab(item.url!!, true, BrowserDirection.FromBookmarks, mode)
+    }
+
+    override fun handleBookmarkDeletion(nodes: Set<BookmarkNode>, eventType: Event) {
+        deleteBookmarkNodes(nodes, eventType)
+    }
+
+    override fun handleBackPressed() {
+        navController.popBackStack()
+    }
+
+    override fun handleSigningIn() {
+        services.launchPairingSignIn(context, navController)
+    }
+
+    private fun openInNewTab(
```

Comment:
is this abstraction necessary? Could we just call openToBrowserAndLoad directly?

Reply:
We could.
Saw this code being used in two methods `handleOpeningBookmark` and `handleBookmarkTapped` and thought that it would be good to extract it in it's own method.
Looking at this class now again maybe we could also extract the code to navigate to a different fragment in a separate private method? There are multiple usages with only the `NavDirections` argument being different.
Not quite accustomed with the project and the coding styles direction, please advise.

Reply:
> Looking at this class now again maybe we could also extract the code to navigate to a different fragment in a separate private method? 

That seems like a great idea, yes.

Reply:
Thanks!
Added that and rebased all commits. Hope everything is good now.